homeassistant/components/bosch_shc/config_flow.py

```diff
@@ -0,0 +1,220 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            try:
+                self.info = info = await self._get_info(host)
+            except SHCConnectionError:
+                errors["base"] = "cannot_connect"
+            except SHCmDNSError:
+                _LOGGER.warning("Error looking up mDNS entry")
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(info["unique_id"])
+                self._abort_if_unique_id_configured({CONF_HOST: host})
+                self.host = host
+                return await self.async_step_credentials()
+
+        return self.async_show_form(
+            step_id="user", data_schema=HOST_SCHEMA, errors=errors
+        )
+
+    async def async_step_credentials(self, user_input=None):
+        """Handle the credentials step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                helper = await self.hass.async_add_executor_job(
+                    SHCRegisterClient, self.host, user_input[CONF_PASSWORD]
+                )
+                result = await self.hass.async_add_executor_job(
+                    helper.register, self.host, "HomeAssistant"
+                )
+                if "token" in result:
+                    self.hostname = result["token"].split(":", 1)[1]
+
+                await self.hass.async_add_executor_job(
+                    write_tls_asset, self.hass, CONF_SHC_CERT, result["cert"]
+                )
+                await self.hass.async_add_executor_job(
+                    write_tls_asset, self.hass, CONF_SHC_KEY, result["key"]
+                )
+                await validate_input(
+                    self.hass,
+                    self.host,
+                    self.hass.config.path(DOMAIN, CONF_SHC_CERT),
+                    self.hass.config.path(DOMAIN, CONF_SHC_KEY),
+                )
+            except SHCAuthenticationError:
+                errors["base"] = "invalid_auth"
+            except SHCConnectionError:
+                errors["base"] = "cannot_connect"
+            except SHCmDNSError:
+                _LOGGER.warning("Error looking up mDNS entry")
+                errors["base"] = "cannot_connect"
+            except SHCSessionError:
+                _LOGGER.warning("API call returned non-OK result. Wrong password?")
+                errors["base"] = "unknown"
+            except SHCRegistrationError:
+                _LOGGER.warning(
+                    "SHC not in pairing mode! Please press the Bosch Smart Home Controller button until LED starts blinking"
+                )
+                errors["base"] = "unknown"
```

Comment:
Why is this an unknown error? We seem to know what the problem is.

Reply:
Bump.

Reply:
I changed the behavior in the 3rd party library: wrong password or bridge not in pairing mode will both result in a registration error `SHCRegistrationError`, and are handled that way.
However, it can happen that session errors occur, when the API will return Non-OK result. This is why I treated them also as `unknown`

Reply:
I introduced a new error type `session_error` for this type of error.